### DISCUSSION TITLE
UDP - Receiver on windows does not use RSS Queues

### DISCUSSION THREAD

**Discussion by Febbe** [esnet/iperf#1297] [Q&A]
# Context

* Version of iperf3: iperf-3.1.3-win64

* Hardware: Intel(R) Ethernet Converged Network Adapter X550-T2

* Operating system (and distribution, if any): Version 1909 (OS Build 18363.959)

_Please note: iperf3 is supported on Linux, FreeBSD, and macOS.
Support may be provided on a best-effort basis to other UNIX-like
platforms.  We cannot provide support for building and/or running
iperf3 on Windows, iOS, or Android. 

Ok gotcha, but where to report this then?

# Bug Report

* Expected Behavior
When running multiple iperf servers, incomming UDP packages are distributed over several RSS Queues, not occupiing only one Prozessor. 

* Actual Behavior
All incomming UDP traffic occupies only one RSS Queue resulting in a fully occupied core. This limits the maximum traffic to ~4 up to ~5 Gbit/s

* Steps to Reproduce
Run on server: 
```bash
start iperf3.exe -s -p 50000
start iperf3.exe -s -p 50001
start iperf3.exe -s -p 50002
```
Run on client:
```bash
start cmd /c "iperf3.exe -c 169.254.234.107 -p 50000 -n 10000000000 -b 0 -M 9014 --udp-counters-64bit -l 32K -u"
start cmd /c "iperf3.exe -c 169.254.234.107 -p 50001 -n 10000000000 -b 0 -M 9014 --udp-counters-64bit -l 32K -u"
start cmd /c "iperf3.exe -c 169.254.234.107 -p 50002 -n 10000000000 -b 0 -M 9014 --udp-counters-64bit -l 32K -u"
```
With tcp and / or on ubuntu linux 20.04 it works:
```bash
start cmd /c "iperf3.exe -c 169.254.234.107 -p 50000 -n 10000000000 -b 0 -M 9014 -l 256K"
start cmd /c "iperf3.exe -c 169.254.234.107 -p 50001 -n 10000000000 -b 0 -M 9014 -l 256K"
start cmd /c "iperf3.exe -c 169.254.234.107 -p 50002 -n 10000000000 -b 0 -M 9014 -l 256K"
```
* Possible Solution

I don't know if it is possible on windows, but try to force the usage of RSS Queues for UDP

**Comment by davidBar-On** [comment]
I am not familiar with Windows RSS queues, but from what I read it seems that this is a system configuration is not controlled by the application.  E.g. see: https://docs.microsoft.com/en-us/windows-hardware/drivers/network/setting-the-number-of-rss-processors

**Comment by Febbe** [comment]
From (Intel NIC X550-T2 Description)[http://www.intel.com/content/www/us/en/embedded/products/networking/ethernet-x550-datasheet.html] i found 
```
The following additional cases are not part of the MSFT RSS specification but are supported RSS modes:
    • UdpIPV4 — The X550 parses the packet to identify a packet with UDP over IPv4.
    • UdpIPV6 — The X550 parses the packet to identify a packet with UDP over IPv6.
```
I am not sure what this means. That Microsoft won't support UDP RSS. Or that it just don't handle it and the NIC will do this automatically.

**Comment by bmah888** [comment]
This is probably not an iperf3 issue, so you probably should not expect to find an answer here.  I think the text means that the NIC supports some modes above and beyond what the Microsoft support requires.  But in either case you would probably need to configure the adapter for whichever RSS mode you want.  My guess is the support for TCP you are seeing is the default behavior for the adapter and you might need to find some utility to set this.

**Comment by tobi-tobsen** [comment]
You can try enabling it as explained here: https://docs.microsoft.com/en-us/windows-hardware/drivers/network/reserving-processors-for-applications

**Comment by Febbe** [comment]
@tobi-tobsen thank you for your hint, but I've already enabled them and I have read tons of documentation and tutorials. Also, they are strangely working for TCP but not for UDP. For me it seems like, that this is a windows issue.

**Comment by Febbe** [comment]
I found out, what's the reason, that the UDP RSS Queues aren't working: The Intel(R) Ethernet Converged Network Adapter X550-T2 does not send Jumbo Packages even if configured (other NIC does), so all UDP packages larger than the default 1500MTU are split into IP packages. IP packages do not have ports and the UDP-RSS-Hashing algorithm can't work. Even if it would know that the IP package is part of an UDP package.

The only way is to reduce the MTU size on the sending side of iperf to the physically MTU.
Also windows does not receive more than one UDP package from the kernel, when using the default WINSOCK API. Which will result in very high CPU load. To solve this iperf should use the WSA Registered I/O API.
